File: LeetCode/All/java/103.binary-tree-zigzag-level-order-traversal.java
// problem: https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode node) {
        if(node==null)
            return new ArrayList<List<Integer>>();
        
        List<List<Integer>> lists = new ArrayList<List<Integer>>();

        Stack<TreeNode> s1 = new Stack<TreeNode>();
        Stack<TreeNode> s2 = new Stack<TreeNode>();

        s1.push(node);
        while(!s1.isEmpty() || !s2.isEmpty()){
            TreeNode current = null;
            ArrayList<Integer> newList = new ArrayList<Integer>();
            while(!s1.isEmpty()){
                 current = s1.pop();
                if(current.left!=null)
                    s2.push(current.left);
                if(current.right!=null)
                    s2.push(current.right);

                // System.out.print(current.data + " ");
                newList.add(current.val);
            }
            
            if(newList.size()!=0)
                lists.add(newList);
            newList = new ArrayList<Integer>();
            
            while(!s2.isEmpty()){
                current = s2.pop();
                if(current.right!=null)
                    s1.push(current.right);
                if(current.left!=null)
                    s1.push(current.left);
                
                // System.out.print(current.data + " ");
                newList.add(current.val);
            }
            
            if(newList.size()!=0)
                lists.add(newList);

        }
        
        return lists;
    }
}

File: LeetCode/Difficulty/Medium/js/103.binary-tree-zigzag-level-order-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var zigzagLevelOrder = function(root) {
    if(!root){
        return [];
    }
    const result = [];
    let queue = [root];
    let deep = 0;
    while(queue.length) {
        let next = [];
        let levels = [];
        while(queue.length){
            const node = queue.shift();
            if(deep%2 === 0){
                levels.push(node.val);
            }
            else {
                levels.unshift(node.val);
            }

            if(node.left)
                next.push(node.left);
            if(node.right)
                next.push(node.right);

        }
        deep++;
        if(levels.length){
            result.push(levels);
        }
        queue = next;
    }
    return result;
};
